chore(slashing): sort payloads by address

just-mitch opened the pull request:
Fix #14436

just-mitch committed:
chore: sort payloads by address

just-mitch commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/14706?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#14706** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/14706?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/14706?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#14684** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/14684?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Maddiaa0 commented on yarn-project/slasher/src/config.ts:
```diff
@@ -31,19 +31,19 @@ export function bigIntToOffence(offense: bigint): Offence {
-  offenses: Offence[];
+  validator: EthAddress;
+  amount: bigint;
+  offense: Offence;
```
spelling is different here

Maddiaa0 commented on yarn-project/slasher/src/slasher_client.ts:
```diff
@@ -208,16 +208,21 @@ export class SlasherClient {
    */
-  private wantToSlash(args: WantToSlashArgs) {
+  private wantToSlash(args: WantToSlashArgs[]) {
     // TODO(#14489): need to sort the payloads by attester address
```
todo can go

just-mitch replied:
Yep, I saw that- it is removed in the PR stacked above.

Maddiaa0 commented on yarn-project/slasher/src/slasher_client.ts:
```diff
@@ -208,16 +208,21 @@ export class SlasherClient {
+  private wantToSlash(args: WantToSlashArgs[]) {
     // TODO(#14489): need to sort the payloads by attester address
-    this.log.info('Wants to slash', args);
+    const sortedArgs = [...args].sort((a, b) => a.validator.toString().localeCompare(b.validator.toString()));
```
is there much point in casting back and forth?

just-mitch replied:
Not sure what you mean? It is dropping `EthAddress` to string for comparison? Unless you're suggesting we just compare directly as bigint or Fr?

Maddiaa0 reviewed (approved):
overall lgtm, some nits

just-mitch committed:
fix: tolerate duplicate proposals (#14732)

We used to only look at payloads that come in through events. 

This was a problem because if the payload was already deployed, we don't
emit an event.

So we need to grab the expected address, and insert it ourselves.

At the same time then, we need to make sure we aren't adding the same
payload to the queue multiple times.

Fix #14694
